File: plugins/fhir/src/handlers/metadata.rs
use axum::extract::{Path, State};
use axum::response::IntoResponse;
use axum::Json;
use std::sync::Arc;

use crate::error::AppError;
use crate::fhir::capability;
use crate::sql_safety::validate_dataset_id;
use crate::state::AppState;

/// Build the SELECT that checks whether a dataset row exists by id in `_datasets`.
pub fn build_dataset_exists_sql(meta_schema: &str, dataset_id: &str) -> String {
    format!(
        "SELECT id FROM {}._datasets WHERE id = '{}'",
        meta_schema,
        dataset_id.replace('\'', "''")
    )
}

pub async fn get_metadata(
    State(state): State<Arc<AppState>>,
    Path(dataset_id): Path<String>,
) -> Result<impl IntoResponse, AppError> {
    validate_dataset_id(&dataset_id)?;

    let check_sql = build_dataset_exists_sql(&state.meta_schema(), &dataset_id);
    let conn = state.new_request_conn().map_err(AppError::Internal)?;
    match conn.execute(check_sql).await {
        crate::query_executor::QueryResult::Select { rows, .. } if !rows.is_empty() => {}
        _ => {
            return Err(AppError::NotFound(format!(
                "Dataset '{}' not found",
                dataset_id
            )));
        }
    }

    let cs = capability::build_capability_statement(
        &state.registry,
        &state.search_params,
        &dataset_id,
    );

    Ok(Json(cs))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn dataset_exists_sql_escapes_quotes() {
        let sql = build_dataset_exists_sql("\"db\".\"meta\"", "d's");
        assert!(sql.starts_with("SELECT id FROM \"db\".\"meta\"._datasets"));
        assert!(sql.contains("'d''s'"));
    }

    #[test]
    fn dataset_exists_sql_plain() {
        let sql = build_dataset_exists_sql("\"m\"", "ds1");
        assert_eq!(sql, "SELECT id FROM \"m\"._datasets WHERE id = 'ds1'");
    }
}

File: plugins/fhir-fn/functions/handlers/metadata.ts
// @ts-nocheck - Deno edge function
// Port of plugins/fhir/src/handlers/metadata.rs

import { Conn } from "../db.ts";
import { FhirError } from "../error.ts";
import { AppState } from "../state.ts";
import { validateDatasetId, toQualifiedMetaSchema } from "../sql_safety.ts";
import { buildCapabilityStatement } from "../fhir/capability.ts";

/**
 * Build the SELECT that checks whether a dataset row exists by id in `_datasets`.
 * Mirrors Rust: format!("SELECT id FROM {}._datasets WHERE id = '{}'", meta_schema, dataset_id.replace('\'', "''"))
 */
export function buildDatasetExistsSql(metaSchema: string, datasetId: string): string {
  return `SELECT id FROM ${metaSchema}._datasets WHERE id = '${datasetId.replaceAll("'", "''")}'`;
}

/**
 * GET /{datasetId}/metadata — return a CapabilityStatement for the dataset.
 * Mirrors Rust get_metadata:
 *   1. validateDatasetId (throws FhirError.badRequest on failure)
 *   2. Check the dataset exists in _datasets; 404 if not
 *   3. Build and return the CapabilityStatement
 */
export async function getMetadata(
  datasetId: string,
  conn: Conn,
  state: AppState,
): Promise<Response> {
  validateDatasetId(datasetId);

  const metaSchema = toQualifiedMetaSchema(state.dbName);
  const sql = buildDatasetExistsSql(metaSchema, datasetId);
  const rows = await conn.query(sql);

  if (rows.length === 0) {
    throw FhirError.notFound(`Dataset '${datasetId}' not found`);
  }

  const cs = buildCapabilityStatement(state.registry, state.searchParams, datasetId);
  return Response.json(cs);
}
